Cache deserialized FormDef by session id

gherceg opened the pull request:
I had to create a new PR since the previous PR was based off of a branch that will no longer be merged. See https://github.com/dimagi/formplayer/pull/1123 for any potentially missing context.

This work is a response to [Clayton's suggestion](https://github.com/dimagi/formplayer/pull/1121#issuecomment-1099297838) to target what specifically we want to optimize. In this case, our main goal is to improve answer timings since a user filling out a form should not have to deserialize a FormDef on every request. This still results in a FormDef being deserialized upon session creation, but that is less likely to experience noticeable gains than answer requests and submission are, since it is an already slower request.

6cd6bc1eb0d9df31a39fc6f17518c4b17cb5f06f is the bulk of this PR, but has a crucial missing component which is that we would not call `getFormDef` until the first answer request, which means we wouldn't be taking advantage of a cached FormDef until the second answer request.

Since we have to deserialize a FormDef upon session creation, it is best to prime the `form_definition` cache with that FormDef which is what 54aff2bfb2d56d7ccbff1c516594115220dbf762 does.

QA: https://dimagi-dev.atlassian.net/browse/QA-4079

snopoke commented on src/main/java/org/commcare/formplayer/services/FormDefinitionService.java:
```diff
@@ -54,4 +54,24 @@ public SerializableFormDefinition getOrCreateFormDefinition(
+     * @param serializedFormDef serialized FormDef object
+     * @return deserialized FormDef object
+     */
+    @Cacheable(key = "#sessionId")
+    public FormDef getFormDef(String sessionId, String serializedFormDef) {
+        FormDef formDef;
+        try {
+            formDef = FormDefStringSerializer.deserialize(serializedFormDef);
```
You could simplify this for callers by taking in the whole session object instead of passing the ID and serialized formDef separately:

```suggestion
    @Cacheable(key = "#session.id")
    public FormDef getFormDef(SerializableFormSession session) {
        FormDef formDef;
        try {
            formDef = FormDefStringSerializer.deserialize(session.getFormDefinition().getSerializedFormDef());
```

gherceg replied:
Good point. See 5068f204cea26929ec844d477ed79d290fb0567c

gherceg commented on src/main/java/org/commcare/formplayer/services/NewFormResponseFactory.java:
```diff
@@ -118,6 +123,9 @@ public NewFormResponse getResponse(FormSession formEntrySession) throws Exceptio
         formSessionService.saveSession(serializedSession);
+        // prime the form def cache for this session
+        Cache cache = this.cacheManager.getCache("form_definition");
+        cache.put(formEntrySession.getSessionId(), formEntrySession.getFormDef());
```
That does simplify things. See bdd1518830166ae3b955e1ed0f7b32a6d561f98a

gherceg commented:
> what's the status of this PR?

On the verge of getting to the bottom of the issue that arose in QA when caching sessions. There's a strange edge case where the question node from the previous request is referenced when answering questions with a null value. Typical behavior is for this question node to be copied as part of the `loadInstanceXml` method, and to set the value on this fresh copy of the question node.

gherceg commented on src/main/java/org/commcare/formplayer/services/FormDefinitionService.java:
```diff
@@ -54,4 +57,34 @@ public SerializableFormDefinition getOrCreateFormDefinition(
+        try {
+            formDef = FormDefStringSerializer.deserialize(session.getFormDefinition().getSerializedFormDef());
+        } catch (Exception e) {
+            formDef = null;
```
That makes sense to me especially with the concerns you brought up in the #1126

gherceg commented:
Finally got to the bottom of the issue. Answers were being set on a node that was cached from the previous answer request. I wasn't aware of this lower level cache so it took a bit of digging. I think the best bet is to always ensure this cache is cleared prior to returning the FormDef from `getFormDef`. I figured it wasn't worth adding a check to determine if the FormDef came from cache or not, since a newly deserialized FormDef doesn't have anything in the cache yet anyway. See a4e94177caee451ada47eba4016fe090821e174f

shubham1g5 commented:
> when writing to local storage it isn't clear how that xmlns value would get mapped to allow us to fetch FormDefs

@gherceg [This is where](https://github.com/dimagi/commcare-core/blob/formplayer/src/main/java/org/javarosa/core/model/FormDef.java#L1671) the xmlns gets mapped.

gherceg commented:
So I took a simple approach to the StorageManager issue and just registered FormDef.STORAGE_KEY as needed in 51c94a657be652c3d2cb0479a6498045352a27af (with a minor change in 6ac2848ca04a1d1ae43d34ef30dbecfba5151b7b). I did this because I noticed a similar pattern in the [CommCarePlatform](https://github.com/dimagi/commcare-core/blob/9a4192a98e640b1fcf0e8b1fc67b13c4f663de96/src/main/java/org/commcare/util/CommCarePlatform.java#L56) which is setup when initializing a `FormSession`. The reason I didn't add it there is because that occurs after we would have already wanted to read the form def from local storage.

Then knowing that the sms entry point did not follow the same pattern as other form entry points with the `ConfigureStorageFromSession` decorator (requires an already existing session), I used the same `configure` line found in the [ConfigureStorageFromSession](https://github.com/dimagi/formplayer/blob/0a6850fdbeb61d8c5dedcda4c4c338e350fbd275/src/main/java/org/commcare/formplayer/aspects/ConfigureStorageFromSessionAspect.java#L29) and added that after the `SerializableFormSession` was created in the `NewFormResponseFactory`. See 07d76fe2a6924089c2b76d0750efa4170cf7e24c.

Lastly, I added a bit of code to ensure the serialized FormDef is updated on the shared form_definition table if deserialization fails in the first place. See 296854b3f1c65fb483fb9624cde9868022225234

snopoke reviewed (commented):
I think it would be better to register the storage preemptively since it makes the code easier to reason about.

There are 2 places I think it should go:
https://github.com/dimagi/formplayer/blob/6cd6bc1eb0d9df31a39fc6f17518c4b17cb5f06f/src/main/java/org/commcare/formplayer/application/AbstractBaseController.java#L92-L95

and

https://github.com/dimagi/formplayer/blob/0783752d088d3f2c34937c722f80c44ae146293a/src/main/java/org/commcare/formplayer/services/NewFormResponseFactory.java#L69-L75

shubham1g5 commented on src/test/java/org/commcare/formplayer/services/FormDefinitionServiceTest.java:
```diff
@@ -162,20 +176,104 @@ public void testGetOrCreateFormDefinitionNewVersionNotEqual() {
+    }
+
+    @Test
+    public void testGetFormDefFromStorageUpdatesSerializableFormDefinition() throws Exception {
```
we can probably abstract out the common setup between this and `testGetFormDefBrokenSerialization` into another method.

gherceg replied:
Agreed that there is room for abstraction, but it's one line that is specifically related to setting up a broken serialization. I also added an additional test that relies on this setup. While all three tests could be consolidated into one test with multiple assertions, in HQ we tend to keep assertions in separate tests to make it easy to reason with.

All that to say I don't mind a bit of repetition if it keeps tests relatively simple. I don't see a super obvious abstraction to be made that wouldn't overcomplicate things, so feel free to highlight that or suggest a change if you want. Otherwise I'll likely leave as is.

snopoke replied:
You could extract out the test setup:

```java
public SerializableFormDefinition createBrokenFormDef() {
    SerializableFormDefinition formDef = this.formDefinitionService.getOrCreateFormDefinition(
            this.appId, this.formXmlns, this.formVersion, this.formDef
    );
    ReflectionTestUtils.setField(formDef, "serializedFormDef", "not a form def");

    // this line added based on PR comments below
    formDefinitionRepo.save(formDef);

    SerializableFormSession session = new SerializableFormSession(UUID.randomUUID().toString());
    session.setFormDefinition(formDef);
    getFormDefStorage().write(this.formDef);
    
    return formDef
}

gherceg replied:
Fair point 😅  see 70111ce3e775fe57f4a4bda134179ab20b554d5d

shubham1g5 commented on src/main/java/org/commcare/formplayer/services/FormDefinitionService.java:
```diff
@@ -54,4 +67,102 @@ public SerializableFormDefinition getOrCreateFormDefinition(
+    }
+
+    private Optional<FormDef> getFormDefFromStorage(String xmlns) {
+        this.storageFactory.getStorageManager().registerStorage(FormDef.STORAGE_KEY, FormDef.class);
```
doesn't `storageFactory` already have the storage registered ?

snopoke replied:
no, that's the whole crux of the matter. Specifically for SMS forms there is no app install and so the storage never get's registered.

gherceg replied:
A `CommCarePlatform` is setup during initialization of a FormSession, but that does not register FormDef storage, only [registers PropertyManager storage](https://github.com/dimagi/commcare-core/blob/9a4192a98e640b1fcf0e8b1fc67b13c4f663de96/src/main/java/org/commcare/util/CommCarePlatform.java#L56). This happens _after_ we need access to a local form def, so we need to register the FormDef storage prior to this point anyway. 

In the context of the FormController, storage is configured due to the `@ConfigureStorageFromSession` decorator used on what appears to be all endpoints except sms. Registering FormDef storage after configuring the factory [here](https://github.com/dimagi/formplayer/blob/master/src/main/java/org/commcare/formplayer/aspects/ConfigureStorageFromSessionAspect.java#L29) makes the most sense to me.

I've been reproducing locally by modifying the FormDefinitionService to attempt to access the local storage on any `getFormDef`, though you really just need to be able to inspect the storage factory at that point. You'll see that the StorageManager returned from the factory does not have any registered storage yet.

gherceg commented:
@snopoke I agree that we should register FormDef storage at a higher level, but related to what I mentioned [here](https://github.com/dimagi/formplayer/pull/1127#discussion_r874827942), 
>  FormDef storage is only registered when setting up a FormplayerConfigEngine, which only happens during navigate menu session requests.

the AbstractBaseController is only used during menu session navigation, so I'm leaning towards adding in the [ConfigureStorageFromSessionAspect](https://github.com/dimagi/formplayer/blob/0a6850fdbeb61d8c5dedcda4c4c338e350fbd275/src/main/java/org/commcare/formplayer/aspects/ConfigureStorageFromSessionAspect.java#L29) instead.

gherceg commented on src/main/java/org/commcare/formplayer/services/NewFormResponseFactory.java:
```diff
@@ -106,6 +105,10 @@ public NewFormResponse getResponse(NewSessionRequestBean bean, String postUrl) t
         );
 
+        // can only setup local storage once a form session has been created
+        this.storageFactory.configure(formSession.getSerializableSession());
```
Totally missed that. I think I got excited cause I tested this locally and it worked, but in hindsight, in order to test sms locally I need to run a menu session first so that localization is properly setup (should be able to put out a fix for that), and I'm wondering if that is the same case here. Is it possible performing an app install for a menu session prior to an sms form session allows the sms form session to have access to a sqlite db?

snopoke replied:
I don't think that would work for SMS since it uses a different data directory so a previous app install wouldn't be accessible

snopoke replied:
For normal sessions a previous app install of the same app for that user would have created the DBs (though each request still needs to register them with the storage manager) and it's quite possible this is only happening after we load the form def from the DB.

ctsims replied:
@gherceg It's quite likely that a previous app install is resulting in a thread later having a Localizer. In Spring the worker threads don't terminate after completion (they are reused between requests). If the issue here is simply that a Localizer was never created on the SMS thread, a previously successful menusession request that created a threadstatic localizer would likely change the behavior for a subsequent sms session request.

gherceg replied:
Yeah I figured the localizer mechanism and the app db mechanisms are quite different. I'm fairly confident what you are describing is exactly what happens in the case of the localizer.

gherceg commented on src/main/java/org/commcare/formplayer/services/FormDefinitionService.java:
```diff
@@ -54,4 +67,102 @@ public SerializableFormDefinition getOrCreateFormDefinition(
+            throw new WrappedException("Error serializing form def", e);
+        }
+        formDefinition.setSerializedFormDef(serializedFormDef);
+        return this.formDefinitionRepo.save(formDefinition);
```
Ah wasn't sure what the java best practices were around the use of `this`. Out of curiosity do we have a formplayer convention? Or is it just use when necessary and that's it?

shubham1g5 replied:
> Or is it just use when necessary and that's it

that's what I follow, I think it's better to omit `this` when not necessary to keep the code less verbose.

snopoke commented on src/main/java/org/commcare/formplayer/services/NewFormResponseFactory.java:
```diff
@@ -71,10 +71,11 @@ public NewFormResponse getResponse(NewSessionRequestBean bean, String postUrl) t
                 bean.getSessionData().getAppId(),
                 bean.getRestoreAs(),
                 bean.getRestoreAsCaseId());
+        storageFactory.getStorageManager().registerStorage(FormDef.STORAGE_KEY, FormDef.class);
```
What do you think about putting this in a method in `FormplayerStorageFactory` with a javadoc to give a bit more context of why it's needed? It would also tie the two usages together.

gherceg replied:
Yeah anything to make this less confusing in the future sounds good to me. See d82f76ea27cde69ac1bd3b1180d744162ebd9307 and let me know if the javadoc could be more informative.

snopoke commented on src/test/java/org/commcare/formplayer/services/FormDefinitionServiceTest.java:
```diff
@@ -162,20 +177,133 @@ public void testGetOrCreateFormDefinitionNewVersionNotEqual() {
+        SerializableFormDefinition updatedFormDef = this.formDefinitionService.getOrCreateFormDefinition(
+                this.appId, this.formXmlns, this.formVersion, this.formDef
+        );
+        assertEquals(updatedFormDef.getSerializedFormDef(),
+                FormDefStringSerializer.serialize(reSerializedFormDef));
```
I don't think this is actually testing anything since the broken form def was never saved to the DB. I think you'd need to save the broken SerializableFormDefinition before calling `formDefinitionService.getFormDef(session)`.

You can confirm this by removing the call to `this.formDefinitionService.getFormDef(session)` and seeing if the test passes.

snopoke replied:
Same goes for `testGetFormDefFromStorageUpdatesCache`

snopoke commented on src/main/java/org/commcare/formplayer/services/FormplayerStorageFactory.java:
```diff
@@ -102,6 +105,15 @@ public void preDestroy() {
     }
 
+    /**
+     * Call after configuring factory if form def storage needs to be accessible
+     * Menu navigation requests setup form def storage as part of MenuSession initialization
+     * Form controller requests do not setup form def storage as part of FormSession initialization
```
nit:
```suggestion
     * Call after configuring factory if form def storage needs to be accessible.
     * Menu navigation requests setup form def storage as part of MenuSession initialization.
     * Form controller requests do not setup form def storage as part of FormSession initialization.
```
